Introduce tropical geometry

lkastner opened the pull request:
I just want to see the documentation for now.

YueRen committed:
new: initial commit Tropical Geometry

lkastner committed:
TropicalGeometry: Fix typo

lkastner committed:
TropicalGeometry: Modify to make run in Oscar

lkastner committed:
TropicalGeometry: Make TropicalHypersurface(poly) work

lkastner committed:
TropicalGeometry: Homogenize inhomogeneous input for polymake

lkastner committed:
TropicalGeomtry: Ask for vertices of tropical hypersurface

YueRen committed:
chg: moving files around, adding basic properties for tropical varieties

YueRen committed:
fix: some broken properties

YueRen committed:
fix: broken property

lkastner committed:
TropicalGeometry: Workaround for minibug in AbstractAlgebra

lkastner committed:
TropicalGeometry: Also homogenize homogeneous polynomials, i.e. add variable

YueRen committed:
fix: tropical numbers

lkastner committed:
TropicalGeometry: Remove brainstorming leftovers

lkastner committed:
TropicalGeometry: Nicer show methods

lkastner committed:
TropicalGeometry: intersect tropical varieties

lkastner committed:
TropicalGeometry: Add documentation to intersect function

lkastner committed:
TropicalGeometry: Elaborate on AbstractAlgebra.jl workaround

lkastner committed:
TropicalGeometry: Install PolyhedralComplex workaround

lkastner committed:
TropicalGeometry: Fix docstring for new Base.show

lkastner committed:
TropicalGeometry: Eliminate most warnings from duplicate definitions

YueRen committed:
chg (broken): restructured codebase

lkastner committed:
TropicalGeometry: Add constructor according to Note in AbstractAlgebra.jl

lkastner committed:
TropicalGeometry: Start documentation

- Add reference to Sturmfels-MacLagan book
- Embed in Algebraic Geometry section

lkastner committed:
TropicalGeometry: Workaround has arrived on master of AbstractAlgebra

lkastner committed:
PolyhedralGeometry: Access to basic properties of PolyhedralComplex

YueRen committed:
chg (broken): constructors for tropical hypersurfaces

YueRen committed:
fix: hypersurface from algebraic polynomial

YueRen committed:
fix: tropical_polynomial coefficients depend on min-max convention

lkastner committed:
TropicalGeometry: Extract underlying poyhedral complex

YueRen committed:
chg: adding more documentation

YueRen committed:
chg: initial changes to base everything on PolyhedralComplex

YueRen committed:
fix: various bugs

lkastner committed:
PolyhedralGeometry: Add some more basic functionality for PolyhedralComplex

YueRen committed:
TropicalGeometry: initial commit for valuation.jl

YueRen committed:
TropicalGeometry: fixing initial forms to work with ValuationMap

YueRen committed:
TropicalGeometry: support for trivial valuations

YueRen committed:
TropicalGeometry: commenting out polyhedral_complex_workaround in main.jl

YueRen committed:
TropicalGeometry: initial commit for initial.jl

YueRen committed:
TropicalGeometry: initial commit for groebner_basis.jl

YueRen committed:
TropicalGeometry: initial (broken) version for tropical groebner bases

YueRen committed:
TropicalGeometry: fixing tropical groebner bases

YueRen committed:
TropicalGeometry: fix groebner_basis header

YueRen committed:
TropicalGeometry: sanatizing input and names in groebner_bases.jl

YueRen committed:
TropicalGeometry: adding complete_reduction parameter to groebner basis

YueRen committed:
TropicalGeometry: using MPolyBuildCtx in pseudo_change_base_ring

YueRen committed:
TropicalGeometry: desimulate_valuation

YueRen committed:
TropicalGeometry: code cleanup

- moved simulate_valuation and desimulate_valuation from groebner_basis.jl to valuation.jl
- merged pseudo_change_base_ring into simulate_valuation

YueRen committed:
TropicalGeometry: added return_lead parameter to groebner_basis

YueRen committed:
TropicalGeometry: initial commit groebner_polyhedron.jl

YueRen committed:
TropicalGeometry: new homogeneity_space

YueRen committed:
TropicalGeometry: code cleanup using Iterators.peel

fingolfin commented on docs/src/PolyhedralGeometry/polyhedral_complexes.md:
```diff
@@ -0,0 +1,54 @@
+$\mathbb{F}^n$, for $n$ fixed, is a *polyhedral complex* if
+
+- the set $\mathcal{F}$ is closed with respect to taking faces and
+- if $C,D\in\mathcal{F}$ then $C\cap D$ is a face of both, $C$ and $D$.
```
```suggestion
- if $C,D\in\mathcal{F}$ then $C\cap D$ is a face of both $C$ and $D$.
```

fingolfin commented on docs/src/TropicalGeometry/intro.md:
````diff
@@ -0,0 +1,29 @@
+## Auxiliary
+
+```@docs
+dual_subdivision(TH::TropicalHypersurface{M, EMB})
````
```suggestion
dual_subdivision
```

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
```diff
@@ -0,0 +1,86 @@
+    pm_complex::Polymake.BigObject
+    function PolyhedralComplex(pm::Polymake.BigObject)
+        return new(pm)
+    end
```
This is redundant

lkastner replied:
No, it is not. I remember clearly that with `PolyhedralFan` there was some issue which made this necessary.

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
```diff
@@ -0,0 +1,86 @@
+###############################################################################
+function Base.show(io::IO, PC::PolyhedralComplex)
+    ad = ambient_dim(PC)
+    if ad == -1.0
```
I am surprised to a see a floating point number here. Is that intentional, and if so, what's the reason?

lkastner replied:
Previously Polymake.jl (and polymake) did not give an error whenever the ambient dimension could not be computed, instead the returned `-1.0`. This has been fixed and I have fixed this in the other PR.

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/properties.jl:
````diff
@@ -0,0 +1,410 @@
+2
+```
+"""
+ambient_dim(PC::PolyhedralComplex) = Polymake.fan.ambient_dim(pm_object(PC))
````
I would recommend to add type assertions to these methods to improve type stability (see  also PR #927). E.g.:
```suggestion
ambient_dim(PC::PolyhedralComplex) = Polymake.fan.ambient_dim(pm_object(PC))::Int
```

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/properties.jl:
```diff
@@ -0,0 +1,410 @@
+
+
+@doc Markdown.doc"""
+    vertices(as, PC)
```
I'd make it clear here that `as` is optional. Also, specify types whenever possible.
```suggestion
    vertices([as::Type,] PC::PolyhedralComplex)
```

fingolfin commented on src/TropicalGeometry/variety_supertype.jl:
````diff
@@ -0,0 +1,590 @@
+# todo: add examples for varieties, curves and linear spaces
+```
+"""
+function codim(T::TropicalVarietySupertype{M,EMB}) where {M,EMB}
````
In places where you don't need to access the type parameters, you can simply omit them. I've just picked this one example, but of course it applies in many other places.
```suggestion
function codim(T::TropicalVarietySupertype)
```

lkastner replied:
Thanks, I will apply this on the new PR.

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
````diff
@@ -0,0 +1,86 @@
+A polyhedral complex in ambient dimension 2
+```
+"""
+function PolyhedralComplex(Incidence::IncidenceMatrix, VR::Union{SubObjectIterator{<:Union{PointVector,PointVector}}, Oscar.MatElem, AbstractMatrix}, far_vertices::Union{Vector{Int}, Nothing} = nothing, L::Union{SubObjectIterator{<:RayVector}, Oscar.MatElem, AbstractMatrix, Nothing} = nothing)
````
Please wrap this across multiple lines. While we don't impose a fixed line width limit, as a rule of thumb, anything that forces me to scroll horizontally when reviewing a PR in the GitHub web interface is bad. If you stick to 80-120 characters per line, that's fine, anything beyond probably isn't.

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
```diff
@@ -0,0 +1,86 @@
+
+
+@doc Markdown.doc"""
+    PolyhedralComplex(Polytopes, VR, far_vertices, Lineality)
```
Is `Lineality` the same as the argument `L` in the actual function declaration? This argument is not mentioned in the description of the arguments either.

fingolfin commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
```diff
@@ -0,0 +1,86 @@
+    PolyhedralComplex(Polytopes, VR, far_vertices, Lineality)
+
+# Arguments
+- `Polytopes::IncidenceMatrix`: An incidence matrix; there is a 1 at position
```
Argument names should be lower case, unless the argument is a type (but then you usually would not do `SomeTypeName::Type` but rather `::Type{SomeTypeName}`).

Same applies to `VR`, too

fingolfin reviewed (commented):
This PR is marked as a draft and also clearly unfinished, so I am somewhat reluctant to comment extensively on it. On the other hand, it is already >2800 lines long, and looks like it'll grow more, and so reviewing it will only become harder unless it gets split somehow.... so I am just leaving some general stylistic comments for which I think it'll save effort on the long run to make them now before you add even more code.

YueRen committed:
TropicalGeometry: initial commit for points.jl, added random_linear_polynomials

lkastner commented:
@fingolfin There is a second PR that will deal with the polyhedral complexes separately. Nevertheless these were needed here, so this branch is based on this.

One question though: Avoiding large PRs in general is good style, but how should one proceed when introducing a new concept? I think the first PR should have all the basic functionality to at least be somewhat useful, even if it becomes large compared to the usual PRs.

YueRen committed:
TropicalGeometry: broken commit so we can continue working on Avi's desktop

YueRen committed:
TropicalGeometry: fixing random_linear_polynomials to return non-homogeous polynomials

YueRen committed:
TropicalGeometry: renamed random_linear_polynomial to random_affine_linear_polynomial

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,80 @@
+end
+export is_homogeneous
+
+function is_homogeneous(I::MPolyIdeal{K} where {K})
```
I was actually surprised to see that this works (but upon re-reading the docs, it's clearly OK), so I learned something, thanks :-)

That said, I am not sure it is useful to write it like this: If you don't reference `K` in the code, then this can (and should) be replaced by
```suggestion
function is_homogeneous(I::MPolyIdeal)
```
If you do plan to reference `K` in the code, this won't work, you'll have to write
```suggestion
function is_homogeneous(I::MPolyIdeal{K}) where {K}
```

fingolfin commented on src/TropicalGeometry/groebner_polyhedron.jl:
```diff
@@ -0,0 +1,48 @@
+I = ideal([x+t*y,y+t*z])
+groebner_polyhedron(I,val_t,w)
+=======#
+function groebner_polyhedron(I,val::ValuationMap{K,p} where {K,p},w::Vector{Int})
```
Also here and in a few more places
```suggestion
function groebner_polyhedron(I, val::ValuationMap, w::Vector{Int})
```

YueRen committed:
TropicalGeometry: first version of tropical_points

YueRen committed:
TropicalGeometry: first version of tropical_link

YueRen committed:
TropicalGeometry: added p_adic_precision to tropical_points

YueRen committed:
TropicalGeometry: changed tropical_link

YueRen committed:
TropicalGeometry: changed tropical_points, increase p_adic_precision if solver unsuccessful

YueRen committed:
TropicalGeometry: changed tropical_links

YueRen committed:
TropicalGeometry: changed tropical_links

YueRen committed:
TropicalGeometry: equipped tropical_points with option remove_points_at_infinity

YueRen committed:
TropicalGeometry: typos and redundant Oscar.

YueRen committed:
TropicalGeometry: initial commit for multiplicity

YueRen committed:
TropicalGeometry: added missing header for multiplicity.jl

YueRen committed:
TropicalGeometry: fixed tropical_points to correctly check for vals of zero entries

YueRen committed:
TropicalGeometry: fixed tropical_link to ignore multiplicities

YueRen committed:
TropicalGeometry: fixed check_weight_vector_for_legality

YueRen committed:
TropicalGeometry: improved multiplicity, allowing for more types of input

YueRen committed:
TropicalGeometry: changed tropical_points to use Avi's new code (currently still broken)

YueRen committed:
TropicalGeometry: code cleanup

YueRen committed:
TropicalGeometr: initial commit for groebner_flip (broken atm)

YueRen committed:
TropicalGeometry: added desimulate_valuation for weight vectors

YueRen committed:
TropicalGeometry: added docu for groebner_basis

YueRen committed:
TropicalGeometry: fixing valuations + initial for ideals

YueRen committed:
TropicalGeometry: added docu for homogeneity_space

YueRen committed:
TropicalGeometry: added docu for tropical_link

YueRen committed:
TropicalGeometry: changed links.jl to link.jl

YueRen committed:
TropicalGeometry: major ValuationMap overhaul

YueRen committed:
TropicalGeometry: added desimulate_valuation for pairs of vectors

YueRen committed:
TropicalGeometry: first version of groebner_flip

YueRen committed:
TropicalGeometry: fixed computing tropical reduced GB

YueRen committed:
TropicalGeometry: fixed tighten_simulation to create distinct x-monomials

YueRen committed:
TropicalGeometry: fixed groebner_basis by introducing t_ecart

YueRen committed:
TropicalGeometry: removed is_homogeneous

YueRen committed:
TropicalGeometry: added show functions for ValuationMap

YueRen committed:
TropicalGeometry: added p_adic_precision option to tropical_points, minor bugfixes

YueRen committed:
TropicalGeometry: tropical_link added checks + restricted input

YueRen committed:
TropicalGeometry: tighten_simulation fixed to divide by gcd of coefficients

YueRen committed:
TropicalGeometry: groebner_basis fixed to use satstd always

YueRen committed:
TropicalGeometry: satstd for coefficient rings workaround

YueRen committed:
TropicalGeometry: tropical_points fixed to be compatible with latest version of pAdicSolver

YueRen committed:
TropicalGeometry: tropical_points fixed output format + added todo

YueRen committed:
TropicalGeometry: added tropical_points for t-adic valuation via p-adic valuation for p>>0

YueRen committed:
TropicalGeometry: initial restructured to allow polys, vectors and ideals as input

YueRen committed:
TropicalGeometry: simulate_valuation restructured to allow for vectors of polys as input

YueRen committed:
TropicalGeometry: link fixed for cases in which zerodimensional cuts are empty

YueRen committed:
TropicalGeometry: interreduce first part

YueRen committed:
TropicalGeometry: allow dot product between fmpq and Int64 vectors

YueRen committed:
TropicalGeometry: traversal initial commit (currently broken)

YueRen committed:
TropicalGeometry: broken commit, checked in for debugging purposes

YueRen committed:
TropicalGeometry: tropical_variety first complete version

YueRen committed:
TropicalGeometry: anchor_point

YueRen committed:
TropicalGeometry: tropical_variety first nontrivial output

YueRen committed:
TropicalGeometry: tropical_variety fixed bug when starting_points on lower-dimensional cells

YueRen committed:
TropicalGeometry: tropical_variety fixed bug which made working_list_done is unsorted

YueRen committed:
TropicalGeometry: tropical_variety first working version

YueRen committed:
TropicalGeometry: tropical_link fixed bug

YueRen committed:
TropicalGeometry: groebner_basis fixed bug caused by typo

YueRen committed:
TropicalGeometry: groebner_polyhedron added skip_reduction optional parameter

YueRen committed:
TropicalGeometry: tropical_variety fixed bug which caused multiple same starting cones

YueRen committed:
TropicalGeometry: interreduce_tropically first complete version

Commit:
TropicalGeometry: added weight for trop hypersurfaces

YueRen committed:
TropicalGeometry: fixing broken merge

YueRen committed:
TropicalGeometry: tropical_points overhaul

YueRen committed:
TropicalGeometry: tropical_variety, check that input valuation is on coefficients of input ideal

YueRen committed:
TropicalGeometry: fixed typo

YueRen committed:
TropicalGeometry: is_valuation_*, making tests more flexible

YueRen committed:
TropicalGeometry: gcd(::Vector{fmpq}) added

YueRen committed:
TropicalGeometry: fixing more typos

YueRen committed:
TropicalGeometry: initial, fixed bug for elements of val 0 but not in val.valued_ring

YueRen committed:
TropicalGeometry: interreduce_tropically fix

YueRen committed:
TropicalGeometry: tropical_link, now allows for ideals over finite fields

YueRen committed:
TropicalGeometry: tropical_variety, added check whether ideal homogeneous

YueRen committed:
TropicalGeometry: tighten_valuation, improved docu

YueRen committed:
TropicalGeometry: tropical_link, fixed bug arising over p-adics (not sure why)

YueRen committed:
TropicalGeometry: tropical_variety, added computation of multiplicity

YueRen committed:
TropicalGeometry: tropical_variety, added homogenization of inhomogeneous ideals

YueRen committed:
TropicalGeometry: fixed typo in comment code

YueRen committed:
TropicalGeometry: added gcd of vectors of fmpq and gfp_elem (workaround)

YueRen committed:
TropicalGeometry: tropical_points, using independent sets for easier cuts to dimension 0

YueRen committed:
TropicalGeometry: t_adic_valuation, extended to more input types

YueRen committed:
TropicalGeometry: tropical_variety, fixed bug in starting point computation

YueRen committed:
TropicalGeometry: tropical_variety, added dehomogenization

YueRen committed:
TropicalGeometry: tropical_variety, fixed bug in homogenize

YueRen committed:
TropicalGeometry: tropical_variety, moved from traversal.jl to variety.jl + storing data in attributes now

YueRen committed:
TropicalGeometry: tropical_numbers, default to min

YueRen committed:
TropicalGeometry: tropical_polynomial, default to 0 and allow passing valuation as argument

YueRen committed:
TropicalGeometry: TropicalHypersurface, default to min and allow passing valuation as argument

YueRen committed:
TropicalGeometry: renamed tropical_numbers to tropical_semiring, overhauled valuation (tropical_variety still bugged with win!)

ValuationMap now is a homomorphism to the tropical_semiring,
so that it can also store the information whether it is the min or max convention.
Changed all functions accordingly

YueRen committed:
TropicalGeometry: initial, forgotten change for last commit

YueRen committed:
TropicalGeometry: Base.:(-)(x::TropicalSemiringElem, y::TropicalSemiringElem, better error message

lkastner commented:
This PR is being split as it grew too large. The first part is #1152 .
